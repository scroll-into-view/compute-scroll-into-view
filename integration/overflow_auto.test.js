beforeAll(async () => {
  await page.goto('http://localhost:3000/integration/overflow_auto')
})

const ScrollLogicalPosition = ['start', 'center', 'end', 'nearest']

describe('scrollMode: always', () => {
  ScrollLogicalPosition.forEach(block => {
    ScrollLogicalPosition.forEach(inline => {
      test(`block: ${block}, inline: ${inline}`, async () => {
        const expected = await page.evaluate(
          options => {
            const container = document.querySelector('.container')
            container.scrollTo(0, 0)
            document.querySelector('.target').scrollIntoView(options)
            const { scrollLeft, scrollTop } = container
            return { left: scrollLeft, top: scrollTop }
          },
          { block, inline }
        )
        const actual = await page.evaluate(
          options => {
            const container = document.querySelector('.container')
            container.scrollTo(0, 0)
            const [{ left, top }] = window.computeScrollIntoView(
              document.querySelector('.target'),
              options
            )
            return { left, top }
          },
          { block, inline }
        )
        expect(expected).toEqual(actual)
        expect(actual).toMatchSnapshot()

        // The 'nearest' cases can have diff behavior depending on the current scroll position
        if (block === 'nearest' || inline === 'nearest') {
          const expected = await page.evaluate(
            options => {
              const container = document.querySelector('.container')
              container.scrollTo(
                container.clientWidth * 3,
                container.clientHeight * 3
              )
              document.querySelector('.target').scrollIntoView(options)
              const { scrollLeft, scrollTop } = container
              return { left: scrollLeft, top: scrollTop }
            },
            { block, inline }
          )
          const actual = await page.evaluate(
            options => {
              const container = document.querySelector('.container')
              container.scrollTo(
                container.clientWidth * 3,
                container.clientHeight * 3
              )
              const [{ left, top }] = window.computeScrollIntoView(
                document.querySelector('.target'),
                options
              )
              return { left, top }
            },
            { block, inline }
          )
          expect(expected).toEqual(actual)
          expect(actual).toMatchSnapshot()
        }
      })
    })
  })
})

describe('scrollMode: if-needed', () => {
  describe('vertical', () => {
    test('completely below the fold', async () => {
      const actual = await page.evaluate(() => {
        const container = document.querySelector('.container')
        container.scrollTo(100, 0)
        return window.computeScrollIntoView(document.querySelector('.target'), {
          scrollMode: 'if-needed',
        })
      })
      expect(actual).toHaveLength(1)
      expect(actual).toMatchSnapshot()
    })

    test('partially below the fold', async () => {
      const actual = await page.evaluate(() => {
        const container = document.querySelector('.container')
        container.scrollTo(100, 50)
        return window.computeScrollIntoView(document.querySelector('.target'), {
          scrollMode: 'if-needed',
        })
      })
      expect(actual).toHaveLength(1)
      expect(actual).toMatchSnapshot()
    })
  })

  describe('horizontal', () => {
    test('completely overflowing', async () => {
      const actual = await page.evaluate(() => {
        const container = document.querySelector('.container')
        container.scrollTo(0, 100)
        return window.computeScrollIntoView(document.querySelector('.target'), {
          scrollMode: 'if-needed',
        })
      })
      expect(actual).toHaveLength(1)
      expect(actual).toMatchSnapshot()
    })

    test('partially overflowing', async () => {
      const actual = await page.evaluate(() => {
        const container = document.querySelector('.container')
        container.scrollTo(50, 100)
        return window.computeScrollIntoView(document.querySelector('.target'), {
          scrollMode: 'if-needed',
        })
      })
      expect(actual).toHaveLength(1)
      expect(actual).toMatchSnapshot()
    })
  })
})
