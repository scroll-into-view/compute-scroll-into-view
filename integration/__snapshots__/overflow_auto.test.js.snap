// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrollMode: always block: center, inline: center 1`] = `
Object {
  "left": 250,
  "top": 250,
}
`;

exports[`scrollMode: always block: center, inline: end 1`] = `
Object {
  "left": 200,
  "top": 250,
}
`;

exports[`scrollMode: always block: center, inline: nearest 1`] = `
Object {
  "left": 200,
  "top": 250,
}
`;

exports[`scrollMode: always block: center, inline: nearest 2`] = `
Object {
  "left": 300,
  "top": 250,
}
`;

exports[`scrollMode: always block: center, inline: start 1`] = `
Object {
  "left": 300,
  "top": 250,
}
`;

exports[`scrollMode: always block: end, inline: center 1`] = `
Object {
  "left": 250,
  "top": 200,
}
`;

exports[`scrollMode: always block: end, inline: end 1`] = `
Object {
  "left": 200,
  "top": 200,
}
`;

exports[`scrollMode: always block: end, inline: nearest 1`] = `
Object {
  "left": 200,
  "top": 200,
}
`;

exports[`scrollMode: always block: end, inline: nearest 2`] = `
Object {
  "left": 300,
  "top": 200,
}
`;

exports[`scrollMode: always block: end, inline: start 1`] = `
Object {
  "left": 300,
  "top": 200,
}
`;

exports[`scrollMode: always block: nearest, inline: center 1`] = `
Object {
  "left": 250,
  "top": 200,
}
`;

exports[`scrollMode: always block: nearest, inline: center 2`] = `
Object {
  "left": 250,
  "top": 300,
}
`;

exports[`scrollMode: always block: nearest, inline: end 1`] = `
Object {
  "left": 200,
  "top": 200,
}
`;

exports[`scrollMode: always block: nearest, inline: end 2`] = `
Object {
  "left": 200,
  "top": 300,
}
`;

exports[`scrollMode: always block: nearest, inline: nearest 1`] = `
Object {
  "left": 200,
  "top": 200,
}
`;

exports[`scrollMode: always block: nearest, inline: nearest 2`] = `
Object {
  "left": 300,
  "top": 300,
}
`;

exports[`scrollMode: always block: nearest, inline: start 1`] = `
Object {
  "left": 300,
  "top": 200,
}
`;

exports[`scrollMode: always block: nearest, inline: start 2`] = `
Object {
  "left": 300,
  "top": 300,
}
`;

exports[`scrollMode: always block: start, inline: center 1`] = `
Object {
  "left": 250,
  "top": 300,
}
`;

exports[`scrollMode: always block: start, inline: end 1`] = `
Object {
  "left": 200,
  "top": 300,
}
`;

exports[`scrollMode: always block: start, inline: nearest 1`] = `
Object {
  "left": 200,
  "top": 300,
}
`;

exports[`scrollMode: always block: start, inline: nearest 2`] = `
Object {
  "left": 300,
  "top": 300,
}
`;

exports[`scrollMode: always block: start, inline: start 1`] = `
Object {
  "left": 300,
  "top": 300,
}
`;

exports[`scrollMode: if-needed horizontal completely overflowing 1`] = `
Array [
  Object {
    "el": Object {},
    "left": 200,
    "top": 250,
  },
]
`;

exports[`scrollMode: if-needed horizontal partially overflowing 1`] = `
Array [
  Object {
    "el": Object {},
    "left": 200,
    "top": 250,
  },
]
`;

exports[`scrollMode: if-needed vertical completely below the fold 1`] = `
Array [
  Object {
    "el": Object {},
    "left": 200,
    "top": 250,
  },
]
`;

exports[`scrollMode: if-needed vertical partially below the fold 1`] = `
Array [
  Object {
    "el": Object {},
    "left": 200,
    "top": 250,
  },
]
`;
